ProcessInstanceAssert::hasVariableWithValue shows inconsistent behaviour with same test case and values

## Description

ProcessInstanceAssert::hasVariableWithValue passes and fails during repeated runs with unchanged test data and code.

### Expected behaviour

Consistent behavior. If object contains expected values then the test should pass (independent of the sequence)

## Reproduction steps

Please run test code repeatedly and observe varying outcome.

<details>
<summary>Test code</summary>

https://github.com/rob2universe/camunda8-testing/blob/a01e62ee443d13ddd38f3cde5fec1a0839109d51/src/test/java/io/camunda/c8/test/ProcessTests.java#L72

```
package io.camunda.c8.test;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import io.camunda.zeebe.client.ZeebeClient;
import io.camunda.zeebe.client.api.response.DeploymentEvent;
import io.camunda.zeebe.process.test.api.ZeebeTestEngine;
import io.camunda.zeebe.process.test.extension.ZeebeProcessTest;
import io.camunda.zeebe.process.test.filters.RecordStream;
import org.junit.jupiter.api.Test;

import java.time.Duration;
import java.util.Map;
import java.util.concurrent.TimeoutException;

import static io.camunda.zeebe.process.test.assertions.BpmnAssert.assertThat;

@ZeebeProcessTest
public class ProcessTests {
  private ZeebeTestEngine engine;
  private ZeebeClient client;
  private RecordStream recordStream;

  private DeploymentEvent initDeployment() {
    return client.newDeployResourceCommand()
        .addResourceFromClasspath("process.bpmn")
        .addResourceFromClasspath("decision.dmn")
        .send()
        .join();
  }

  @Test
  public void testDeployment() {
    assertThat(initDeployment());
  }

  @Test
  public void testProcess() throws InterruptedException, TimeoutException {
    initDeployment();

    // When instance is started
    var piEvent = client.newCreateInstanceCommand()
        .bpmnProcessId("TestProcess")
        .latestVersion()
        .variables(Map.of("myItem", "a"))
        .send()
        .join();
    // Then instance should have passed start event and should be awaiting job completion
    assertThat(piEvent)
        .hasPassedElement("ProcessingStartedStartEvent")
        .isWaitingAtElements("CallServiceTask");

    // When job is activated
    var response = client.newActivateJobsCommand()
        .jobType("callService")
        .maxJobsToActivate(1)
        .send()
        .join();
    // Then activated job should exist
    var activatedJob = response.getJobs().get(0);
    assertThat(activatedJob);

    // When job is completed and process engine had time to continue processing
    client.newCompleteCommand(activatedJob.getKey()).send().join();
    engine.waitForIdleState(Duration.ofMillis(500));

    // Then service task, business rule task, and process instance should be completed
    assertThat(piEvent)
        .hasPassedElementsInOrder("CallServiceTask", "EvaluateBusinessRulesTask")
        .isCompleted()
        // and business rule task result should be available as process data
        .hasVariableWithValue("result", Map.of("checkedItem","a","myOutput","aa"));
    //TODO
    // the test passed and sometimes fails, probably depending on the sequence of the serialization of the Map.
    // When the test failes the error is:
    // java.lang.AssertionError: The variable 'result' does not have the expected value. The value passed in
    // ('{checkedItem=a, myOutput=aa}') is internally mapped to a JSON String that yields
    // '{"checkedItem":"a","myOutput":"aa"}'. However, the actual value (as JSON String) is
    // '{"myOutput":"aa","checkedItem":"a"}'.

  }
}
```

</details>

<details>
<summary>Process</summary>

```xml

<?xml version="1.0" encoding="UTF-8"?>
<definitions xmlns="https://www.omg.org/spec/DMN/20191111/MODEL/" xmlns:dmndi="https://www.omg.org/spec/DMN/20191111/DMNDI/" xmlns:dc="http://www.omg.org/spec/DMN/20180521/DC/" xmlns:modeler="http://camunda.org/schema/modeler/1.0" xmlns:camunda="http://camunda.org/schema/1.0/dmn" id="Definitions_99d90b2e-3675-44a8-842b-7e38de90a7b4" name="CheckItem" namespace="http://camunda.org/schema/1.0/dmn" exporter="Camunda Modeler" exporterVersion="5.0.0" modeler:executionPlatform="Camunda Cloud" modeler:executionPlatformVersion="1.3.0" camunda:diagramRelationId="2ffb3ea9-7596-432a-a9ce-1bac47a013a4">
  <decision id="Decision_CheckItem" name="Check Item">
    <decisionTable id="DecisionTable_0azx1hv">
      <input id="Input_1" label="myItem">
        <inputExpression id="InputExpression_1" typeRef="string">
          <text>myItem</text>
        </inputExpression>
      </input>
      <output id="OutputClause_1gt1jlu" label="myOutput" name="myOutput" typeRef="string" />
      <output id="OutputClause_15hqo4g" label="checkedItem" name="checkedItem" typeRef="string" />
      <rule id="DecisionRule_0qmyhsl">
        <inputEntry id="UnaryTests_1hk0huc">
          <text>"a"</text>
        </inputEntry>
        <outputEntry id="LiteralExpression_0ofzcke">
          <text>"aa"</text>
        </outputEntry>
        <outputEntry id="LiteralExpression_18inzt7">
          <text>myItem</text>
        </outputEntry>
      </rule>
      <rule id="DecisionRule_05d3x9i">
        <inputEntry id="UnaryTests_1rbftn4">
          <text>"b"</text>
        </inputEntry>
        <outputEntry id="LiteralExpression_0jzgss5">
          <text>"bb"</text>
        </outputEntry>
        <outputEntry id="LiteralExpression_1gqywzr">
          <text>myItem</text>
        </outputEntry>
      </rule>
      <rule id="DecisionRule_0ozdf9j">
        <inputEntry id="UnaryTests_087nmtg">
          <text>"c"</text>
        </inputEntry>
        <outputEntry id="LiteralExpression_0yeghgd">
          <text>"cc"</text>
        </outputEntry>
        <outputEntry id="LiteralExpression_1bhx39z">
          <text>myItem</text>
        </outputEntry>
      </rule>
    </decisionTable>
  </decision>
  <dmndi:DMNDI>
    <dmndi:DMNDiagram>
      <dmndi:DMNShape dmnElementRef="Decision_CheckItem">
        <dc:Bounds height="80" width="180" x="160" y="100" />
      </dmndi:DMNShape>
    </dmndi:DMNDiagram>
  </dmndi:DMNDI>
</definitions>
```
```xml
<?xml version="1.0" encoding="UTF-8"?>
<bpmn:definitions xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:zeebe="http://camunda.org/schema/zeebe/1.0" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:modeler="http://camunda.org/schema/modeler/1.0" id="Definitions_1fvxzlk" targetNamespace="http://bpmn.io/schema/bpmn" exporter="Camunda Modeler" exporterVersion="5.0.0" modeler:executionPlatform="Camunda Cloud" modeler:executionPlatformVersion="1.1.0">
  <bpmn:process id="TestProcess" name="Test Process" isExecutable="true">
    <bpmn:startEvent id="ProcessingStartedStartEvent" name="Processing started">
      <bpmn:outgoing>Flow_006yvg0</bpmn:outgoing>
    </bpmn:startEvent>
    <bpmn:sequenceFlow id="Flow_006yvg0" sourceRef="ProcessingStartedStartEvent" targetRef="CallServiceTask" />
    <bpmn:endEvent id="ProcessingCompletedEndEvent" name="Processing completed">
      <bpmn:incoming>Flow_18z712h</bpmn:incoming>
    </bpmn:endEvent>
    <bpmn:sequenceFlow id="Flow_08flgx2" sourceRef="CallServiceTask" targetRef="EvaluateBusinessRulesTask" />
    <bpmn:serviceTask id="CallServiceTask" name="Call Service">
      <bpmn:extensionElements>
        <zeebe:taskDefinition type="callService" />
      </bpmn:extensionElements>
      <bpmn:incoming>Flow_006yvg0</bpmn:incoming>
      <bpmn:outgoing>Flow_08flgx2</bpmn:outgoing>
    </bpmn:serviceTask>
    <bpmn:sequenceFlow id="Flow_18z712h" sourceRef="EvaluateBusinessRulesTask" targetRef="ProcessingCompletedEndEvent" />
    <bpmn:businessRuleTask id="EvaluateBusinessRulesTask" name="Evaluate business rules">
      <bpmn:extensionElements>
        <zeebe:calledDecision decisionId="Decision_CheckItem" resultVariable="result" />
      </bpmn:extensionElements>
      <bpmn:incoming>Flow_08flgx2</bpmn:incoming>
      <bpmn:outgoing>Flow_18z712h</bpmn:outgoing>
    </bpmn:businessRuleTask>
  </bpmn:process>
  <bpmndi:BPMNDiagram id="BPMNDiagram_1">
    <bpmndi:BPMNPlane id="BPMNPlane_1" bpmnElement="TestProcess">
      <bpmndi:BPMNEdge id="Flow_08flgx2_di" bpmnElement="Flow_08flgx2">
        <di:waypoint x="340" y="120" />
        <di:waypoint x="400" y="120" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_006yvg0_di" bpmnElement="Flow_006yvg0">
        <di:waypoint x="188" y="120" />
        <di:waypoint x="240" y="120" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_18z712h_di" bpmnElement="Flow_18z712h">
        <di:waypoint x="500" y="120" />
        <di:waypoint x="562" y="120" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id="Event_13xfe7s_di" bpmnElement="ProcessingStartedStartEvent">
        <dc:Bounds x="152" y="102" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="143" y="145" width="55" height="27" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_0lsf20p_di" bpmnElement="CallServiceTask">
        <dc:Bounds x="240" y="80" width="100" height="80" />
        <bpmndi:BPMNLabel />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_0p2r0n9_di" bpmnElement="ProcessingCompletedEndEvent">
        <dc:Bounds x="562" y="102" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="553" y="145" width="55" height="27" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_030jwac_di" bpmnElement="EvaluateBusinessRulesTask">
        <dc:Bounds x="400" y="80" width="100" height="80" />
        <bpmndi:BPMNLabel />
      </bpmndi:BPMNShape>
    </bpmndi:BPMNPlane>
  </bpmndi:BPMNDiagram>
</bpmn:definitions>

```

</details>

## Environment

- OS: <!-- e.g. Linux -->
- Version: <!-- e.g. 1.0.0 -->

--- a/assertions/pom.xml
+++ b/assertions/pom.xml
@@ -29,11 +29,6 @@
 
     <dependency>
       <groupId>io.camunda</groupId>
-      <artifactId>zeebe-test-util</artifactId>
-    </dependency>
-
-    <dependency>
-      <groupId>io.camunda</groupId>
       <artifactId>zeebe-process-test-filters</artifactId>
     </dependency>
 
@@ -45,6 +40,16 @@
     <dependency>
       <groupId>io.camunda</groupId>
       <artifactId>zeebe-client-java</artifactId>
+    </dependency>
+
+    <dependency>
+      <groupId>com.fasterxml.jackson.core</groupId>
+      <artifactId>jackson-core</artifactId>
+    </dependency>
+
+    <dependency>
+      <groupId>com.fasterxml.jackson.core</groupId>
+      <artifactId>jackson-databind</artifactId>
     </dependency>
 
     <!-- Test scope-->

--- a/assertions/pom.xml
+++ b/assertions/pom.xml
@@ -29,6 +29,11 @@
 
     <dependency>
       <groupId>io.camunda</groupId>
+      <artifactId>zeebe-test-util</artifactId>
+    </dependency>
+
+    <dependency>
+      <groupId>io.camunda</groupId>
       <artifactId>zeebe-process-test-filters</artifactId>
     </dependency>
 

--- a/assertions/src/main/java/io/camunda/zeebe/process/test/assertions/ProcessInstanceAssert.java
+++ b/assertions/src/main/java/io/camunda/zeebe/process/test/assertions/ProcessInstanceAssert.java
@@ -17,7 +17,6 @@ package io.camunda.zeebe.process.test.assertions;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
-import io.camunda.zeebe.client.impl.ZeebeObjectMapper;
 import io.camunda.zeebe.process.test.filters.IncidentRecordStreamFilter;
 import io.camunda.zeebe.process.test.filters.ProcessInstanceRecordStreamFilter;
 import io.camunda.zeebe.process.test.filters.RecordStream;
@@ -478,28 +477,7 @@ public class ProcessInstanceAssert extends AbstractAssert<ProcessInstanceAssert,
    * @return this ${@link ProcessInstanceAssert}
    */
   public ProcessInstanceAssert hasVariable(final String name) {
-    final Map<String, String> variables = getProcessInstanceVariables();
-    return assertVariableInMapOfVariables(name, variables);
-  }
-
-  /**
-   * Assert that the given variable name is a key in the given map of variables.
-   *
-   * <p>This assertion has been extracted from the method ${@link #hasVariable(String)} so that the
-   * method ${@link #hasVariableWithValue(String, Object)} could reuse it without having to traverse
-   * the record stream to collect the variables a second time.
-   *
-   * @param name The name of the variable
-   * @param variables The map of variables
-   * @return this ${@link ProcessInstanceAssert}
-   */
-  private ProcessInstanceAssert assertVariableInMapOfVariables(
-      final String name, final Map<String, String> variables) {
-    assertThat(variables)
-        .withFailMessage(
-            "Process with key %s does not contain variable with name `%s`. Available variables are: %s",
-            actual, name, variables.keySet())
-        .containsKey(name);
+    getVariableAssert().containsVariable(name);
     return this;
   }
 
@@ -511,18 +489,7 @@ public class ProcessInstanceAssert extends AbstractAssert<ProcessInstanceAssert,
    * @return this ${@link ProcessInstanceAssert}
    */
   public ProcessInstanceAssert hasVariableWithValue(final String name, final Object value) {
-    final ZeebeObjectMapper mapper = new ZeebeObjectMapper();
-    final String mappedValue = mapper.toJson(value);
-    final Map<String, String> variables = getProcessInstanceVariables();
-
-    assertVariableInMapOfVariables(name, variables);
-    assertThat(variables)
-        .withFailMessage(
-            "The variable '%s' does not have the expected value. The value passed in"
-                + " ('%s') is internally mapped to a JSON String that yields '%s'. However, the "
-                + "actual value (as JSON String) is '%s'.",
-            name, value, mappedValue, variables.get(name))
-        .containsEntry(name, mappedValue);
+    getVariableAssert().hasVariableWithValue(name, value);
 
     return this;
   }
@@ -545,6 +512,10 @@ public class ProcessInstanceAssert extends AbstractAssert<ProcessInstanceAssert,
                 VariableRecordValue::getName,
                 VariableRecordValue::getValue,
                 (value1, value2) -> value2));
+  }
+
+  private VariablesMapAssert getVariableAssert() {
+    return new VariablesMapAssert(getProcessInstanceVariables());
   }
 
   /**

--- a/assertions/src/main/java/io/camunda/zeebe/process/test/assertions/VariablesMapAssert.java
+++ b/assertions/src/main/java/io/camunda/zeebe/process/test/assertions/VariablesMapAssert.java
@@ -1,0 +1,72 @@
+/*
+ * Copyright © 2021 camunda services GmbH (info@camunda.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.camunda.zeebe.process.test.assertions;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.fail;
+
+import io.camunda.zeebe.client.impl.ZeebeObjectMapper;
+import io.camunda.zeebe.test.util.JsonUtil;
+import java.util.Map;
+import org.assertj.core.api.AbstractAssert;
+
+/**
+ * Assertion for variable maps {@code Map<String, String>} which contain JSON strings as values
+ * (e.g. variables in process instances, messages, job activations)
+ */
+public class VariablesMapAssert extends AbstractAssert<VariablesMapAssert, Map<String, String>> {
+
+  private static final ZeebeObjectMapper OBJECT_MAPPER = new ZeebeObjectMapper();
+
+  public VariablesMapAssert(final Map<String, String> actual) {
+    super(actual, VariablesMapAssert.class);
+  }
+
+  /**
+   * Assert that the given variable name is a key in the given map of variables. *
+   *
+   * @param name The name of the variable
+   * @return this ${@link ProcessInstanceAssert}
+   */
+  public VariablesMapAssert containsVariable(final String name) {
+    assertThat(actual)
+        .withFailMessage(
+            "Unable to find variable with name `%s`. Available variables are: %s",
+            name, actual.keySet())
+        .containsKey(name);
+    return this;
+  }
+
+  public VariablesMapAssert hasVariableWithValue(final String name, final Object value) {
+    final String expectedValue = OBJECT_MAPPER.toJson(value);
+    final String actualValue = actual.get(name);
+
+    containsVariable(name);
+
+    try {
+      JsonUtil.assertEquality(expectedValue, actualValue);
+    } catch (final AssertionError e) {
+      fail(
+          "The variable '%s' does not have the expected value. The value passed in"
+              + " ('%s') is internally mapped to a JSON String that yields '%s'. However, the "
+              + "actual value (as JSON String) is '%s'.",
+          name, value, expectedValue, actualValue);
+    }
+
+    return this;
+  }
+}

--- a/assertions/src/main/java/io/camunda/zeebe/process/test/assertions/VariablesMapAssert.java
+++ b/assertions/src/main/java/io/camunda/zeebe/process/test/assertions/VariablesMapAssert.java
@@ -17,10 +17,12 @@
 package io.camunda.zeebe.process.test.assertions;
 
 import static org.assertj.core.api.Assertions.assertThat;
-import static org.assertj.core.api.Assertions.fail;
 
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
 import io.camunda.zeebe.client.impl.ZeebeObjectMapper;
-import io.camunda.zeebe.test.util.JsonUtil;
+import java.io.IOException;
 import java.util.Map;
 import org.assertj.core.api.AbstractAssert;
 
@@ -30,7 +32,12 @@ import org.assertj.core.api.AbstractAssert;
  */
 public class VariablesMapAssert extends AbstractAssert<VariablesMapAssert, Map<String, String>> {
 
+  static final ObjectMapper JSON_MAPPER = new ObjectMapper();
   private static final ZeebeObjectMapper OBJECT_MAPPER = new ZeebeObjectMapper();
+
+  static {
+    JSON_MAPPER.configure(JsonParser.Feature.ALLOW_SINGLE_QUOTES, true);
+  }
 
   public VariablesMapAssert(final Map<String, String> actual) {
     super(actual, VariablesMapAssert.class);
@@ -52,21 +59,31 @@ public class VariablesMapAssert extends AbstractAssert<VariablesMapAssert, Map<S
   }
 
   public VariablesMapAssert hasVariableWithValue(final String name, final Object value) {
+    containsVariable(name);
+
     final String expectedValue = OBJECT_MAPPER.toJson(value);
     final String actualValue = actual.get(name);
 
-    containsVariable(name);
-
-    try {
-      JsonUtil.assertEquality(expectedValue, actualValue);
-    } catch (final AssertionError e) {
-      fail(
-          "The variable '%s' does not have the expected value. The value passed in"
-              + " ('%s') is internally mapped to a JSON String that yields '%s'. However, the "
-              + "actual value (as JSON String) is '%s'.",
-          name, value, expectedValue, actualValue);
-    }
+    assertThat(isEqual(actualValue, expectedValue))
+        .withFailMessage(
+            "The variable '%s' does not have the expected value. The value passed in"
+                + " ('%s') is internally mapped to a JSON String that yields '%s'. However, the "
+                + "actual value (as JSON String) is '%s'.",
+            name, value, expectedValue, actualValue)
+        .isTrue();
 
     return this;
+  }
+
+  private static boolean isEqual(final String actualJson, final String expectedJson) {
+    return asJsonNode(actualJson).equals(asJsonNode(expectedJson));
+  }
+
+  private static JsonNode asJsonNode(final String json) {
+    try {
+      return JSON_MAPPER.readTree(json);
+    } catch (final IOException e) {
+      throw new RuntimeException(e);
+    }
   }
 }

--- a/assertions/src/test/java/io/camunda/zeebe/process/test/assertions/VariablesMapAssertTest.java
+++ b/assertions/src/test/java/io/camunda/zeebe/process/test/assertions/VariablesMapAssertTest.java
@@ -1,0 +1,129 @@
+/*
+ * Copyright © 2021 camunda services GmbH (info@camunda.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.camunda.zeebe.process.test.assertions;
+
+import static org.assertj.core.api.Assertions.assertThatNoException;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+class VariablesMapAssertTest {
+
+  private static final String KEY = "key";
+  private static final String UNKNOWN_KEY = "unknownKey";
+
+  private static final Map<String, String> VARIABLES = new HashMap<>();
+
+  static {
+    VARIABLES.put(KEY, "\"value\"");
+  }
+
+  @Nested
+  class ContainsVariable {
+
+    @Test
+    void shouldNotThrowAssertionErrorWhenVariableIsKnown() {
+      // given
+      final VariablesMapAssert sut = new VariablesMapAssert(VARIABLES);
+
+      // when + then
+      assertThatNoException().isThrownBy(() -> sut.containsVariable(KEY));
+    }
+
+    @Test
+    void shouldThrowAssertionErrorWhenVariableIsUnknown() {
+      // given
+      final VariablesMapAssert sut = new VariablesMapAssert(VARIABLES);
+
+      // when + then
+      assertThatThrownBy(() -> sut.containsVariable(UNKNOWN_KEY))
+          .isExactlyInstanceOf(AssertionError.class)
+          .hasMessage(
+              "Unable to find variable with name `%s`. Available variables are: %s",
+              UNKNOWN_KEY, VARIABLES.keySet());
+    }
+  }
+
+  @Nested
+  class HasVariableWithValue {
+
+    @Test
+    void shouldNotThrowAssertionErrorWhenVariableIsKnownAndValueIsCorrect() {
+      // given
+      final VariablesMapAssert sut = new VariablesMapAssert(VARIABLES);
+
+      // when + then
+      assertThatNoException().isThrownBy(() -> sut.hasVariableWithValue(KEY, "value"));
+    }
+
+    @Test
+    void shouldThrowAssertionErrorWhenVariableIsUnknown() {
+      // given
+      final VariablesMapAssert sut = new VariablesMapAssert(VARIABLES);
+
+      // when + then
+      assertThatThrownBy(() -> sut.hasVariableWithValue(UNKNOWN_KEY, null))
+          .isExactlyInstanceOf(AssertionError.class)
+          .hasMessage(
+              "Unable to find variable with name `%s`. Available variables are: %s",
+              UNKNOWN_KEY, VARIABLES.keySet());
+    }
+
+    @Test
+    void shouldThrowAssertionErrorWhenValueIsDifferent() {
+      // given
+      final VariablesMapAssert sut = new VariablesMapAssert(VARIABLES);
+
+      // when + then
+      assertThatThrownBy(() -> sut.hasVariableWithValue(KEY, "someOtherValue"))
+          .isExactlyInstanceOf(AssertionError.class)
+          .hasMessage(
+              "The variable 'key' does not have the expected value. The value passed in ('someOtherValue') is internally mapped to a JSON String that yields '\"someOtherValue\"'. However, the actual value (as JSON String) is '\"value\"'.");
+    }
+
+    /**
+     * Verifies that the logic to compare the actual and expected value is lenient with respect to
+     * the order of fields in a JSON object
+     */
+    @Test
+    void shouldBeLenientWithRespectToFieldOrder() {
+      // given
+      final Map<String, Boolean> value = new HashMap<>();
+      value.put("a", true);
+      value.put("b", false);
+
+      final String jsonRepresentation1 = "{ \"a\" : true, \"b\": false }";
+      final String jsonRepresentation2 = "{ \"b\" : false, \"a\": true }";
+
+      final Map<String, String> map1 = new HashMap<>();
+      map1.put(KEY, jsonRepresentation1);
+
+      final Map<String, String> map2 = new HashMap<>();
+      map2.put(KEY, jsonRepresentation2);
+
+      final VariablesMapAssert sut1 = new VariablesMapAssert(map1);
+      final VariablesMapAssert sut2 = new VariablesMapAssert(map2);
+
+      // when + then
+      assertThatNoException().isThrownBy(() -> sut1.hasVariableWithValue(KEY, value));
+      assertThatNoException().isThrownBy(() -> sut2.hasVariableWithValue(KEY, value));
+    }
+  }
+}

--- a/assertions/src/test/java/io/camunda/zeebe/process/test/assertions/VariablesMapAssertTest.java
+++ b/assertions/src/test/java/io/camunda/zeebe/process/test/assertions/VariablesMapAssertTest.java
@@ -93,7 +93,7 @@ class VariablesMapAssertTest {
 
       // when + then
       assertThatThrownBy(() -> sut.hasVariableWithValue(KEY, "someOtherValue"))
-          .isExactlyInstanceOf(AssertionError.class)
+          .isInstanceOf(AssertionError.class)
           .hasMessage(
               "The variable 'key' does not have the expected value. The value passed in ('someOtherValue') is internally mapped to a JSON String that yields '\"someOtherValue\"'. However, the actual value (as JSON String) is '\"value\"'.");
     }

--- a/qa/abstracts/src/main/java/io/camunda/zeebe/process/test/qa/abstracts/assertions/AbstractProcessInstanceAssertTest.java
+++ b/qa/abstracts/src/main/java/io/camunda/zeebe/process/test/qa/abstracts/assertions/AbstractProcessInstanceAssertTest.java
@@ -15,6 +15,7 @@
  */
 package io.camunda.zeebe.process.test.qa.abstracts.assertions;
 
+import static org.assertj.core.api.Assertions.assertThatNoException;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
@@ -36,7 +37,6 @@ import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.TimeoutException;
 import org.assertj.core.api.Assertions;
-import org.assertj.core.api.SoftAssertions;
 import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
 
@@ -408,12 +408,12 @@ public abstract class AbstractProcessInstanceAssertTest {
               engine, client, ProcessPackLoopingServiceTask.PROCESS_ID, TYPED_TEST_VARIABLES);
 
       // then
-      final SoftAssertions softly = new SoftAssertions();
-
-      TYPED_TEST_VARIABLES.forEach(
-          (key, value) -> BpmnAssert.assertThat(instanceEvent).hasVariableWithValue(key, value));
-
-      softly.assertAll();
+      assertThatNoException()
+          .isThrownBy(
+              () ->
+                  TYPED_TEST_VARIABLES.forEach(
+                      (key, value) ->
+                          BpmnAssert.assertThat(instanceEvent).hasVariableWithValue(key, value)));
     }
 
     @Test
@@ -998,8 +998,8 @@ public abstract class AbstractProcessInstanceAssertTest {
       assertThatThrownBy(() -> BpmnAssert.assertThat(instanceEvent).hasVariable(expectedVariable))
           .isInstanceOf(AssertionError.class)
           .hasMessage(
-              "Process with key %s does not contain variable with name `%s`. Available variables are: [%s]",
-              instanceEvent.getProcessInstanceKey(), expectedVariable, actualVariable);
+              "Unable to find variable with name `%s`. Available variables are: [%s]",
+              expectedVariable, actualVariable);
     }
 
     @Test

